### PR TITLE
Replace custody challenge game with JABS

### PR DESCRIPTION
**Pull request by vbuterin**
Replace the existing proof of custody game with a new game ("Justin's Awesome Bit Sum" or JABS) that works as follows:

* The data `D` is split up into 512-byte chunks `D[0] .... D[n-1]`, and use a mix function `mix(subkey, data) -> {0,1}` (currently the first bit of the hash of `subkey+data`). We calculate `M[i] = (mix(D[0]) + ... + mix(D[i-1])) % 2`, and set the custody bit to `M[n-1]`
* Anyone can challenge by providing the full `M` where `M[n-1]` is not equal to the custody bit
* Anyone can respond to a challenge by providing a specific position in `M` along with a branch of the data where `M[i-1] ^ mix(D[i]) != M[i]`

The maximum size of data is now `2**6` epochs *  `2**6` blocks * `2**14` bytes = `2**26` bytes, so assuming 512-byte mix chunks the maximum mix size is `2**17` bits or `2**14` bytes. The average mix size is `2**8` bytes.

### CODE REVIEW

**Comment by JustinDrake** [comment]
Ready for an initial review :)

Note that the honest validator behaviour has been moved to a validator doc in #818.

File: specs/core/1_shard-data-chains.md [hunk 1]
```diff
@@ -501,395 +252,484 @@ Define a `SubkeyReveal` as follows:
...
-    # Responder wins
-    if hash(challenge.responder_subkey + response.data) == challenge.current_custody_tree_node:
-        penalize_validator(state, challenge.challenger_index, challenge.responder_index)
-    # Challenger wins
+
+    if is_early_reveal:
+        assert state.validator_registry[reveal.validator_index].slashed_epoch > get_current_epoch(state)
+        slash_validator(state, reveal.validator_index, reveal.revealer_index)
+        state.validator_balances[reveal.revealer_index] += base_reward(state, index) // MINOR_REWARD_QUOTIENT
+    elif reveal.period == state.validator_registry[reveal.validator_index].next_subkey_to_reveal and reveal.mask == ZERO_HASH:
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]: By the variable names, it seems "Subkey reveals" section only processes `early_subkey_reveals`; are they all in `is_early_reveal` cases? Or it's better to adjust the variable naming?

File: specs/core/1_shard-data-chains.md [hunk 2]
```diff
@@ -501,395 +252,484 @@ Define a `SubkeyReveal` as follows:
...
-    # Otherwise just check the response against the root
-    else:
-        assert root == challenge_data.current_custody_tree_node
-    # Update challenge data
-    challenge.deadline=FAR_FUTURE_EPOCH
-    responder.withdrawable_epoch = get_current_epoch(state) + MAX_POC_RESPONSE_DEPTH
-```
+def process_branch_response(state: BeaconState,
+                            response: BranchResponse) -> None:
+    data_challenge = get_data_challenge_record(response.challenge_id)
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]:
```suggestion
    data_challenge = get_data_challenge_record(state, response.challenge_id)
```

File: specs/core/1_shard-data-chains.md [hunk 3]
```diff
@@ -501,395 +252,484 @@ Define a `SubkeyReveal` as follows:
...
+    if data_challenge is not None:
+        return process_data_challenge_response(state, response, data_challenge)
 
-#### Interactive custody challenge continuations
+    custody_challenge = get_custody_challenge_record(response.challenge_id)
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]:
```suggestion
    custody_challenge = get_custody_challenge_record(state, response.challenge_id)
```

File: specs/core/1_shard-data-chains.md [hunk 4]
```diff
@@ -501,395 +252,484 @@ Define a `SubkeyReveal` as follows:
...
+    # Case 2: punitive early reveal
+    # Masking prevents proposer stealing the whistleblower reward
+    # Secure under the aggregate extraction infeasibility assumption
+    # See pages 11-12 of https://crypto.stanford.edu/~dabo/pubs/papers/aggreg.pdf
     else:
-        pubkeys=[pubkey, mask_pubkey]
-        message_hashes=[hash(int_to_bytes8(period)), mask]
-        
-    return bls_multi_verify(
+        pubkeys.append(state.validator_registry[reveal.revealer_index].pubkey)
```
[the remark this answers is not in the document]
Comment (JustinDrake) [reply specs/core/1_shard-data-chains.md]: That was a bug, thanks! We're using bls_verify_multiple in the case there's masking.

**Pull request merged**